file_store.go

```diff
@@ -0,0 +1,286 @@
+func (fs *FileStore) Get(_ context.Context, serverAddress string) (auth.Credential, error) {
+	fs.contentRWLock.RLock()
+	defer fs.contentRWLock.RUnlock()
+
+	authCfg, ok := fs.getAuthConfig(serverAddress)
+	if !ok {
+		return auth.EmptyCredential, nil
+	}
+
+	cred := auth.Credential{
+		Username:     authCfg.Username,
+		Password:     authCfg.Password,
+		RefreshToken: authCfg.IdentityToken,
+		AccessToken:  authCfg.RegistryToken,
+	}
+	if authCfg.Auth != "" {
+		var err error
+		// override username and password
+		cred.Username, cred.Password, err = decodeAuth(authCfg.Auth)
+		if err != nil {
+			return auth.EmptyCredential, fmt.Errorf("failed to decode username and password: %w: %v", ErrInvalidConfigFormat, err)
+		}
+	}
+	return cred, nil
+}
+
+// Put saves credentials into the store for the given server address.
+// Returns ErrPlainTextSaveDisabled if fs.DisableSave is set to true.
+func (fs *FileStore) Put(_ context.Context, serverAddress string, cred auth.Credential) error {
+	if fs.DisableSave {
+		return ErrPlainTextSaveDisabled
+	}
+
+	fs.contentRWLock.Lock()
+	defer fs.contentRWLock.Unlock()
+
+	fs.updateAuths(serverAddress, cred)
+	return fs.saveFile()
+}
+
+// Delete removes credentials from the store for the given server address.
+func (fs *FileStore) Delete(ctx context.Context, serverAddress string) error {
+	fs.contentRWLock.Lock()
+	defer fs.contentRWLock.Unlock()
+
+	if _, err := os.Stat(fs.configPath); os.IsNotExist(err) {
+		// no ops if the config file does not exist
+		return nil
+	}
+	authsMap, ok := fs.content[configFieldAuthConfigs].(map[string]interface{})
+	if !ok {
+		// no ops
+		return nil
+	}
+	if _, ok = authsMap[serverAddress]; !ok {
+		// no ops
+		return nil
+	}
+
+	// update data
+	delete(authsMap, serverAddress)
+	fs.content[configFieldAuthConfigs] = authsMap
+	return fs.saveFile()
+}
+
+// getAuthConfig reads the config and returns authConfig for serverAddress.
+func (fs *FileStore) getAuthConfig(serverAddress string) (authConfig, bool) {
+	authsMap, ok := fs.content[configFieldAuthConfigs].(map[string]interface{})
+	if !ok {
+		return authConfig{}, false
+	}
+	authConfigObj, ok := authsMap[serverAddress].(map[string]interface{})
+	if !ok {
+		return authConfig{}, false
+	}
+
+	var authCfg authConfig
+	for k, v := range authConfigObj {
+		switch k {
+		case configFieldUsername:
+			authCfg.Username, _ = v.(string)
+		case configFieldPassword:
+			authCfg.Password, _ = v.(string)
+		case configFieldBasicAuth:
+			authCfg.Auth, _ = v.(string)
+		case configFieldIdentityToken:
+			authCfg.IdentityToken, _ = v.(string)
+		case configFieldRegistryToken:
+			authCfg.RegistryToken, _ = v.(string)
+		}
+	}
+	return authCfg, true
+}
+
+// updateAuths updates the Auths field of fs.content based on cred.
+func (fs *FileStore) updateAuths(serverAddress string, cred auth.Credential) {
+	authsMap, ok := fs.content[configFieldAuthConfigs].(map[string]interface{})
```

Comment:
do we need to use a lock?

Reply:
We lock the whole put operation at [line 138](https://github.com/oras-project/oras-credentials-go/pull/43/files#diff-61cdeba4f566237d4a3a9c1b205f8fdf1424170eaccd77924d8ae766825394caR138).